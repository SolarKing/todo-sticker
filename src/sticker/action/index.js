// @flow

export class TodoDescriptionChangedIPC {
    id: string;
    description: string;
}

export type CloseTodoNoteIpcAction = {
    id: string;
}