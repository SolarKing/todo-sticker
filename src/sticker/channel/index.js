// @flow
export const IpcChannels = {
    noteDescriptionSend: 'noteDescriptionSend',
    todoDescriptionChanged: 'todoDescriptionChanged',
    closeTodoNote: 'closeTodoNote'
};