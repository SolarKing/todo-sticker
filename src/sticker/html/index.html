<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css" integrity="sha384-fnmOCqbTlWIlj8LyTjo7mOUStjsKC4pOpQbqyi7RrhN7udi9RwhKkMHpvLbHG9Sr" crossorigin="anonymous">
	<title>React Setup</title>
</head>
<body>
	<!-- 欲插入 React Component 的位置 -->
    <div id="app"></div>
    <!-- <script type="text/javascript" src="index_bundle.js"></script> -->
</body>
</html>
