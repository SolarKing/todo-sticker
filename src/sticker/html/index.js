// @flow
import React from 'react';
import ReactDOM from 'react-dom';
import StickerDescription from '../../component/stickerDescription';
import { injectGlobal } from 'styled-components';
import StickerToolBar from '../../component/stickerToolbar';

const app = document.getElementById('app');

function renderToolBarIfOnWindows() {
    console.log('process.platform', process.platform);
    
    if (/^win/i.test(process.platform)) {
        return (
            <StickerToolBar />
        );
    }
}

export const Sticker = () => {
    return (
        <div>
            {renderToolBarIfOnWindows()}
            <StickerDescription />
        </div>
    );
};

app && ReactDOM.render(
    <Sticker />,
    app
);

injectGlobal`
    body {
        width: 100vw;
        height: 100vh;
        margin: 0;
        background-color: #f7f7a3;
        overflow-wrap: break-word;
        overflow-x: hidden;
        -webkit-user-select: none;
        -webkit-app-region: drag;
    }
`;


