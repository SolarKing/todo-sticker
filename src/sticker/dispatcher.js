// @flow
import { type StickerActions } from './action';
import { ipcRenderer } from 'electron';

class Dispatcher {
    dispatch(action: StickerActions) {
        ipcRenderer.send(action.channel, action);
    }
}
const stickerDispatcher = new Dispatcher();

export default stickerDispatcher;