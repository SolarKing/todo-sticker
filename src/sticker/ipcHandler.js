// @flow
import { ipcMain, BrowserWindow, Electron } from 'electron';
import { IpcChannels } from './channel';
import type { CreateStickerAction, EditStickerAction, DeleteStickerAction } from './action';

declare var __dirname: any;
declare var process: any;

class StickerHandler {

    notes: Map<string, BrowserWindow>;
    _createNoteAndSendDescription: Function;
    _updateTodoNoteDescription: Function;
    _closeTodoNote: Function;
    _windowOptions: any = {
        width: 350, 
        height: 300, 
        frame: false, 
        titleBarStyle: 'hidden',
        minWidth: 180,
        minHeight: 150,
    }

    constructor() {
        this.notes = new Map();
        this._createNoteAndSendDescription = this._createNoteAndSendDescription.bind(this);
        this._updateTodoNoteDescription = this._updateTodoNoteDescription.bind(this);
        this._closeTodoNote = this._closeTodoNote.bind(this);
    }

    registerListener() {
        ipcMain.on(IpcChannels.createSticker, this._createNoteAndSendDescription);
        ipcMain.on(IpcChannels.editSticker, this._updateTodoNoteDescription);
        ipcMain.on(IpcChannels.deleteSticker, this._closeTodoNote);
    }

    _createNoteAndSendDescription(event: Electron.event, action: CreateStickerAction) {

        if (this.notes.has(action.id) && this.notes.get(action.id)) {
            let win: BrowserWindow = this.notes.get(action.id);
            win.focus();
            return;
        }

        let win: BrowserWindow = new BrowserWindow(this._windowOptions);

        win.loadURL(`file://${__dirname}/note.html`);

        if (process.env.NODE_ENV === 'development') {
            win.webContents.openDevTools();
        }

        win.webContents.on('dom-ready', () => {
            win.webContents.send(IpcChannels.editSticker, action.description);
        });
        this.notes.set(action.id, win);
        win.on('closed', () => this._removeTodoNotes(action.id));
    }

    _updateTodoNoteDescription(event: Electron.event, action: EditStickerAction) {
        let id = action.id;
        let win: BrowserWindow = this.notes.get(id);
        if (!win) {
            this.notes.delete(id);
            return;
        }
        win.webContents.send(action.channel, action.description);
    }

    _closeTodoNote(event: Electron.event, action: DeleteStickerAction) {
        let win: BrowserWindow = this.notes.get(action.id);
        if (!win) {
            this.notes.delete(action.id);
            return;
        }
        
        win.destroy();
    }

    _removeTodoNotes(id: string) {
        if (!this.notes.has(id)) {
            return;
        }
        this.notes.delete(id);
    }
}

const stickerHandler = new StickerHandler();
export default stickerHandler;