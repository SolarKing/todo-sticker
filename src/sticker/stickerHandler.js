// @flow
import { ipcMain, BrowserWindow, Electron } from 'electron';
import { IpcChannels } from './channel';
import type { CreateStickerAction, EditStickerAction, DeleteStickerAction } from './action';

declare var __dirname: any;
declare var process: any;

class StickerHandler {

    stickers: Map<string, BrowserWindow>;
    _windowOptions: any = {
        width: 350, 
        height: 300, 
        frame: false, 
        titleBarStyle: 'hidden',
        minWidth: 180,
        minHeight: 150,
    }

    constructor() {
        this.stickers = new Map();
    }

    registerListener() {
        ipcMain.on(IpcChannels.createSticker, this._createStickerAndSendEdit.bind(this));
        ipcMain.on(IpcChannels.editSticker, this._editSticker.bind(this));
        ipcMain.on(IpcChannels.deleteSticker, this._deleteSticker.bind(this));
    }

    _createStickerAndSendEdit(event: Electron.event, action: CreateStickerAction) {

        if (this.stickers.has(action.id) && this.stickers.get(action.id)) {
            let win: BrowserWindow = this.stickers.get(action.id);
            win.focus();
            return;
        }

        let win: BrowserWindow = new BrowserWindow(this._windowOptions);

        win.loadURL(`file://${__dirname}/sticker.html`);

        if (process.env.NODE_ENV === 'development') {
            win.webContents.openDevTools();
        }

        win.webContents.on('dom-ready', () => {
            win.webContents.send(IpcChannels.editSticker, action.description);
        });
        this.stickers.set(action.id, win);
        win.on('closed', () => this._removeStickersById(action.id));
    }

    _editSticker(event: Electron.event, action: EditStickerAction) {
        let id = action.id;
        let win: BrowserWindow = this.stickers.get(id);
        if (!win) {
            this.stickers.delete(id);
            return;
        }
        win.webContents.send(action.channel, action.description);
    }

    _deleteSticker(event: Electron.event, action: DeleteStickerAction) {
        let win: BrowserWindow = this.stickers.get(action.id);
        if (!win) {
            this.stickers.delete(action.id);
            return;
        }
        
        win.destroy();
    }

    _removeStickersById(id: string) {
        if (!this.stickers.has(id)) {
            return;
        }
        this.stickers.delete(id);
    }
}

const stickerHandler = new StickerHandler();
export default stickerHandler;