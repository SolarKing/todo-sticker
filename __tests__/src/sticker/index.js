// @flow
import React from 'react';
import { shallow } from 'enzyme';
import { Sticker } from '../../../src/sticker/html';

describe('StickerToolbar', () => {
    let originalPlatform;

    beforeAll(() => {
        originalPlatform = process.platform;
    });

    afterAll(() => {
        Object.defineProperty(process, 'platform', {  
            value: originalPlatform
        });
    });

    it('should render correct when platform is windows', () => {
        Object.defineProperty(process, 'platform', {  
            value: 'win32'
        });
        expect(shallow(<Sticker/>)).toMatchSnapshot();
    });

    it('should not render when platform is mac', () => {
        Object.defineProperty(process, 'platform', {  
            value: 'darwin'
        });
        expect(shallow(<Sticker/>)).toMatchSnapshot();
    });

});