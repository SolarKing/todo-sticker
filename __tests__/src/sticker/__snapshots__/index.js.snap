// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickerToolbar should not render when platform is mac 1`] = `
<div>
  <StickerDescription />
</div>
`;

exports[`StickerToolbar should render correct when platform is windows 1`] = `
<div>
  <StickerToolBar />
  <StickerDescription />
</div>
`;
